engine/src/main/java/org/pentaho/di/trans/steps/sql/ExecSQL.java

```diff
@@ -0,0 +1,332 @@
+          Object valueData = row[index];
+
+          // replace the '?' with the String in the row.
+          //
+          int pos = data.markerPositions.get( i );
+          String replaceValue = valueMeta.getString( valueData );
+          replaceValue = Const.NVL( replaceValue, "" );
+          if ( meta.isQuoteString() && ( valueMeta.getType() == ValueMetaInterface.TYPE_STRING ) ) {
+            // Have the database dialect do the quoting.
+            // This also adds the quotes around the string
+            replaceValue = meta.getDatabaseMeta().quoteSQLString( replaceValue );
+          }
+          buf.replace( pos, pos + 1, replaceValue );
+        }
+        sql = buf.toString();
+      } else {
+        sql = data.sql;
+      }
+    }
+    if ( log.isRowLevel() ) {
+      logRowlevel( BaseMessages.getString( PKG, "ExecSQL.Log.ExecutingSQLScript" ) + Const.CR + sql );
+    }
+
+    boolean sendToErrorRow = false;
+    String errorMessage = null;
+    try {
+
+      if ( meta.isSingleStatement() ) {
+        data.result = data.db.execStatement( sql, data.paramsMeta, paramsData );
+      } else {
+        data.result = data.db.execStatements( sql, data.paramsMeta, paramsData );
+      }
+
+      RowMetaAndData add =
+        getResultRow( data.result, meta.getUpdateField(), meta.getInsertField(), meta.getDeleteField(), meta
+          .getReadField() );
+
+      row = RowDataUtil.addRowData( row, getInputRowMeta().size(), add.getData() );
+
+      if ( !data.db.isAutoCommit() ) {
+        data.db.commit();
+      }
+
+      putRow( data.outputRowMeta, row ); // send it out!
+
+      if ( checkFeedback( getLinesWritten() ) ) {
+        if ( log.isBasic() ) {
+          logBasic( BaseMessages.getString( PKG, "ExecSQL.Log.LineNumber" ) + getLinesWritten() );
+        }
+      }
+    } catch ( KettleException e ) {
+      if ( getStepMeta().isDoingErrorHandling() ) {
+        sendToErrorRow = true;
+        errorMessage = e.toString();
+      } else {
+        throw new KettleStepException( BaseMessages.getString( PKG, "ExecSQL.Log.ErrorInStep" ), e );
+      }
+
+      if ( sendToErrorRow ) {
+        // Simply add this row to the error row
+        putError( getInputRowMeta(), row, 1, errorMessage, null, "ExecSQL001" );
+      }
+    }
+    return true;
+  }
+
+  @Override
+  public void dispose( StepMetaInterface smi, StepDataInterface sdi ) {
+    meta = (ExecSQLMeta) smi;
+    data = (ExecSQLData) sdi;
+
+    if ( log.isBasic() ) {
+      logBasic( BaseMessages.getString( PKG, "ExecSQL.Log.FinishingReadingQuery" ) );
+    }
+
+    if ( data.db != null ) {
+      data.db.disconnect();
+    }
+
+    super.dispose( smi, sdi );
+  }
+
+  /** Stop the running query */
+  @Override
+  public void stopRunning( StepMetaInterface smi, StepDataInterface sdi ) throws KettleException {
+    meta = (ExecSQLMeta) smi;
+    data = (ExecSQLData) sdi;
+
+    if ( data.db != null && !data.isCanceled ) {
+      synchronized ( data.db ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Object data.db synchronized on line 270 should not be re-assigned in the same synchronized block on line 271. After this re-assignment, you lose the reference to the synchronized object.